terraform/main.tf

```diff
@@ -0,0 +1,290 @@
+  # checkov:skip=CKV_GIT_4: We are passing the secret from the random_password resource which is sensitive by default
+  # and not checking in any plain text values. State is always secured.
+
+  for_each = var.organization_secrets
+
+  plaintext_value = random_password.this[each.key].result
+  secret_name     = each.key
+  visibility      = each.value.visibility
+}
+
+# Github Branch Protection Resource
+# https://registry.terraform.io/providers/integrations/github/latest/docs/resources/branch_protection
+#
+# resource "github_branch_protection" "this" {
+#
+#   # GitHub pull requests should require at least 2 approvals
+#   # checkov:skip=CKV_GIT_5: 1 approval is reasonable for a small team
+#   for_each = local.branch_protections
+#
+#   enforce_admins                  = false
+#   pattern                         = "main"
+#   repository_id                   = github_repository.this[each.key].name
+#   require_conversation_resolution = true
+#   required_linear_history         = true
+#   require_signed_commits          = true
+#
+#   required_pull_request_reviews {
+#     dismiss_stale_reviews           = true
+#     require_code_owner_reviews      = true
+#     required_approving_review_count = 1
+#   }
+#
+#   required_status_checks {
+#     contexts = each.value.required_status_checks_contexts
+#     strict   = true
+#   }
+#
+#   restrict_pushes {
+#     push_allowances = each.value.push_allowances
+#   }
+# }
+
+# GitHub Membership Resource
+# https://registry.terraform.io/providers/integrations/github/latest/docs/resources/membership
+
+resource "github_membership" "this" {
+  for_each = local.users
+
+  role     = each.value
+  username = each.key
+}
+
+# Github Organization Security Manager Resource
+# https://registry.terraform.io/providers/integrations/github/latest/docs/resources/organization_security_manager
+
+resource "github_organization_security_manager" "this" {
+  team_slug = github_team.parents["security-team"].slug
+}
+
+# Github Repository Resource
+# https://registry.terraform.io/providers/integrations/github/latest/docs/resources/repository
+
+resource "github_repository" "this" {
+
+  # Ensure GitHub repository is Private
+  # checkov:skip=CKV_GIT_1: Public is ok for us since we are an open source project
+
+  for_each = var.repositories
+
+  allow_auto_merge            = each.value.allow_auto_merge
+  allow_merge_commit          = each.value.allow_merge_commit
+  allow_rebase_merge          = each.value.allow_rebase_merge
+  allow_squash_merge          = each.value.allow_squash_merge
+  allow_update_branch         = each.value.allow_update_branch
+  archive_on_destroy          = true
+  delete_branch_on_merge      = true
+  description                 = each.value.description
+  has_downloads               = each.value.has_downloads
+  has_discussions             = each.value.has_discussions
+  has_issues                  = true
+  has_projects                = true
+  has_wiki                    = false
```

Comment:
@cunla Some beginner questions: Does `has_wiki = false` (vs something like `has_wiki = each.value.has_wiki` mean we are stating that the repos in the django-commons org are forced to not have wikis? Is this something the repos could override if they wanted to? (I'm not attached to the wiki, I just picked it as an example.)

Reply:
Good question. I should have elaborated more on the PR.

So file roles:
- `variables.tf` - define variable types (classes?), notice there is `variable "repositories" {...` there which has a few variables marked as optional with default values. Why I chose to have `has_discussions` as a repo variable while `has_issues` as a constant - I am embarrassed to say I don't have a better answer than laziness :smile: - I just figured if this is the path we want to take, we can continue adding to it.
- `production.tfvars` - instances, should strictly follow the types in `variables.tf`.
- `main.tf` - build configuration based on instances values from `production.tfvars` (or, if not defined explicitly, then default value from `variables.tf`)

As I wrote in the email, pretty much all repository settings can be configured as constants or variables (with/without a default value).

To give some ideas on how we might adopt it:
We can define our "desired/default" repository configuration, and within this configuration:
- What is enforced from day one (i.e., constant in `resource "github_repository" "this"`)
- What is recommended but can be changed by users (i.e., variable with a default value that can be updated in `production.tfvars`) => Note this can also help us review outliers, you can see all repos which have non-default values in the `prod.tfvars` file.
- What is determined by users (i.e., variables without default value, like `description`) 
- What is not configured in the infra-as-code (currently, for example, repo-labels).